### ISSUE TITLE
Change Request: Stricter rule test validation

### ISSUE THREAD

**Issue by bmish** [eslint/eslint#15104]
### ESLint version

8.0.0-rc.0

### What problem do you want to solve?

* Currently, it's easy to forget to assert what suggestions a test case yields, or that a test case yields no assertions. As a result, this can be an area of poor test coverage, leading to bugs around suggestions.
* It's also inconvenient to have to add `suggestions: []` to every single invalid test case for a rule or tests that do not produce suggestions, in order to achieve the desired test coverage.

### What do you think is the correct solution?

Invalid test cases that yield suggestions should require a `suggestions` assertion.

This would be the same as how invalid test cases that produce an autofix require an `output` assertion, and fail with this error:

```
AssertionError [ERR_ASSERTION]: The rule fixed the code. Please add 'output' property.
```

This behavior was added as a [breaking change](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class) in ESLint v7.

We should add a similar assertion for suggestions.

```
AssertionError [ERR_ASSERTION]: The rule violation produced a suggestion. Please add a 'suggestions' property to the error object.
```

Example of invalid test case with suggestions assertion:

```js
new RuleTester().run('example-rule', rule, {
  valid: [],
  invalid: [
    {
      code: `foo();`,
      output: null,
      errors: [
        {
          messageId: 'bar',
          type: 'Property',
          suggestions: [], // or array of suggestions if rule provides them
        },
      ],
    },
  ]
});
```

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by bmish** [comment]
Would this require an RFC? If so, I can put that together so that it can be lined up for v9.

Also, have we begun tracking breaking changes for v9 yet? I'd like to make sure this gets on the list.

**Comment by bmish** [comment]
Still planning to do this, assigned myself.

**Comment by bmish** [comment]
I have opened an RFC for this issue, along with some related rule test validation improvements: https://github.com/eslint/rfcs/pull/84

**Comment by fisker** [comment]
👎 for this.

In some rule, I only add one test for the `suggestions` is enough, they focus other thets on the auto-fix or something else, eg "message", "report loaction", why I need assert `suggestions` in all those?

The tester should not force anything, test one thing in one case is good practice!

**Comment by nzakas** [comment]
Please review the RFC and leave your comments there.

**Comment by bmish** [comment]
Reopening since the RFC hasn't been implemented yet.

**Comment by mdjermanovic** [comment]
Since the related https://github.com/eslint/rfcs/pull/84 includes more changes than just requiring suggestion assertions, I changed the title of this issue so that it tracks all the changes, i.e. the full implementation of https://github.com/eslint/rfcs/pull/84:

> We will add additional assertions to the rule test runner to address several potential points of poor test coverage, especially related to [rule suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).
> 
> Assertions added to invalid test case error objects:
> 
> - Must contain `suggestions` if the test case produces suggestions
> - Must contain the violation message (`message` or `messageId`, but not both)
> 
> Assertions added to invalid test case suggestion objects:
> 
> - Must contain the suggestion code `output`
> - Must contain the suggestion message (`messageId` or `desc`, but not both)
> - The suggestion code `output` must differ from the test case `code`
> 
> Assertions added to all test cases:
> 
> - Any autofix `output` must differ from the test case `code`
> - Cannot have identical test cases
> - The optional `filename` property must be a string
> - The optional `only` property must be a boolean
